packages/core-app-api/src/apis/implementations/SignalsApi/SignalsClient.ts

```diff
@@ -0,0 +1,164 @@
+ */
+export class SignalsClient implements SignalsApi {
+  private endpoint: string;
+  private ws?: Socket;
+  private signalsEnabled: boolean;
+  private subscriptions: Map<string, SignalsSubscription>;
+  private readonly messageQueue: Set<{ channel: string; message: any }>;
+  private readonly identity: IdentityApi;
+
+  static create(options: { configApi: ConfigApi; identityApi: IdentityApi }) {
+    const { configApi, identityApi } = options;
+    return new SignalsClient(configApi, identityApi);
+  }
+
+  private constructor(config: ConfigApi, identity: IdentityApi) {
+    this.subscriptions = new Map();
+    this.messageQueue = new Set();
+    this.signalsEnabled = false;
+    this.endpoint = '';
+    this.identity = identity;
+    this.readConfig(config);
+  }
+
+  private readConfig(config: ConfigApi) {
+    const baseUrl =
+      config.getOptionalString('backend.baseUrl') ?? 'http://localhost/';
+    const ws = config.getOptional('backend.signals');
+    if (ws === true) {
+      this.signalsEnabled = true;
+      this.endpoint = baseUrl;
+      return;
+    }
+
+    this.signalsEnabled =
+      config.getOptionalBoolean('backend.signals.enabled') ?? false;
+    this.endpoint =
+      config.getOptionalString('backend.signals.endpoint') ?? baseUrl;
+  }
+
+  private async connect() {
+    const { token } = await this.identity.getCredentials();
+
+    if (!this.ws) {
+      this.ws = io(this.endpoint, {
+        path: '/signals',
+        auth: { token: token },
+        multiplex: true,
+      });
+    } else if (!this.ws.connected) {
+      this.ws.connect();
+    }
+
+    this.ws.on('connect', () => {
+      for (const msg of this.messageQueue) {
+        this.send(msg.channel, msg.message);
+      }
+      this.messageQueue.clear();
+    });
+
+    this.ws.on('message', (msg: SignalsMessage) => {
+      for (const subscription of this.subscriptions.values()) {
+        if (
+          subscription.pluginId === msg.pluginId &&
+          (!msg.topic || subscription.topic === msg.topic)
+        ) {
+          try {
+            subscription.onMessage(msg.data);
+          } catch (_) {
+            // ignore callback errors
+          }
+        }
+      }
+    });
+  }
+
+  private async send(channel: string, message: any) {
+    if (!this.ws) {
+      await this.connect();
+      this.messageQueue.add({ channel, message });
+      return;
+    }
+
+    try {
+      this.ws.emit(channel, message);
+    } catch (_e) {
+      this.messageQueue.add({ channel, message });
+    }
+  }
+
+  private getSubscriptionKey = (pluginId: string, topic?: string) => {
+    return topic ? `${pluginId}:${topic}` : pluginId;
+  };
+
+  async subscribe(
+    pluginId: string,
+    onMessage: (data: unknown) => void,
+    topic?: string,
+  ) {
+    if (!this.signalsEnabled) {
+      return;
+    }
+
+    const subscriptionKey = this.getSubscriptionKey(pluginId, topic);
+    if (this.subscriptions.has(subscriptionKey)) {
+      throw new Error(
```

Comment:
At the moment, single client can have only one subscription to plugin (& topic). Is the apiRef same for all frontend components? What happens if two components want to subscribe to same data at the same time?